File: backend/src/main/java/hu/bsstudio/raktr/dto/scannable/ScannableDetailsDto.java
package hu.bsstudio.raktr.dto.scannable;

import hu.bsstudio.raktr.dto.user.UserSummaryDto;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.OffsetDateTime;

@Data
@NoArgsConstructor
@AllArgsConstructor
public class ScannableDetailsDto {

    private Long id;

    private String assetTag;

    private String barcode;

    private String name;

    private Integer weight;

    private boolean publicRentable;

    private boolean deleted;

    private String category;

    private String location;

    private Owner owner;

    private OffsetDateTime createdAt;

    private UserSummaryDto createdBy;

    private OffsetDateTime updatedAt;

    private UserSummaryDto updatedBy;

    @Data
    public static class Owner {

        private Long id;

        private String name;

    }

}

File: frontend/src/app/model/scannable/scannableDetailsDto.ts
import {Owner} from '../owner/owner';
import {UserDetails} from '../user/userDetails';

export class ScannableDetailsDto {
  id: number;
  assetTag: string;
  barcode: string;
  name: string;
  weight: number;
  publicRentable: boolean;
  deleted: boolean;
  category: string;
  location: string;
  owner: Owner;
  createdAt: Date;
  createdBy: UserDetails;
  updatedAt: Date;
  updatedBy: UserDetails;

  constructor(
    id: number,
    assetTag: string,
    barcode: string,
    name: string,
    weight: number,
    publicRentable: boolean,
    deleted: boolean,
    category: string,
    location: string,
    owner: Owner,
    createdAt: Date,
    createdBy: UserDetails,
    updatedAt: Date,
    updatedBy: UserDetails
  ) {
    this.id = id;
    this.assetTag = assetTag;
    this.barcode = barcode;
    this.name = name;
    this.weight = weight;
    this.publicRentable = publicRentable;
    this.deleted = deleted;
    this.category = category;
    this.location = location;
    this.owner = owner;
    this.createdAt = createdAt;
    this.createdBy = createdBy;
    this.updatedAt = updatedAt;
    this.updatedBy = updatedBy;
  }

  static fromJson(json: Record<string, unknown>): ScannableDetailsDto {
    return new ScannableDetailsDto(
      json['id'] as number,
      json['assetTag'] as string,
      json['barcode'] as string,
      json['name'] as string,
      json['weight'] as number,
      json['publicRentable'] as boolean,
      json['deleted'] as boolean,
      json['category'] as string,
      json['location'] as string,
      json['owner'] as Owner,
      new Date(json['createdAt'] as string),
      UserDetails.fromJson(json['createdBy'] as Record<string, unknown>),
      new Date(json['updatedAt'] as string),
      UserDetails.fromJson(json['updatedBy'] as Record<string, unknown>)
    );
  }
}
